src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameLexer.cs

```diff
@@ -40,7 +41,7 @@ internal Token GetNext()
         //
         internal Token GetNext(out string tokenString)
         {
-            tokenString = null;
+            tokenString = string.Empty;
             while (char.IsWhiteSpace(_chars[_index]))
```

Comment:
Since this code is so concerned with preserving corner-case behaviors ... I think this will accept more whitespace characters than the unmanaged parser.

Reply:
(FWIW, I do not think that the way this is trying to preserve handling of zeros inside the string is useful. I would be perfectly happy with declaring 0 to be invalid inside assembly names.)

Reply:
TBH, I did not question the implementation much assuming we have shipped it already at least once and any deviation was deemed acceptable.
I assume the native implementation is basically the spec for the assembly name "language" which we want to match, within reason. 
I'll check the discrepancies.

Reply:
Ther is also '\u' kind of escaping that managed lexer misses. That results in `ParseUnicode()`

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/binder/inc/stringlexer.inl#L170

Reply:
There is also support for UTF16 surrogate pairs, but it seems only used when reading digits of `\u` kind of escaped character. 

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/binder/inc/stringlexer.inl#L207

All other uses of `DecodeUTF16Character()` are commented out. I wonder - was it unfinished or disabled later?

Reply:
I think `\u` escaping should be supported and added to managed implementation, but `DecodeUTF16Character()` should be dropped on the native side.

Reply:
Managed parser is missing support for `language` and `custom` attributes.

Reply:
> (FWIW, I do not think that the way this is trying to preserve handling of zeros inside the string is useful. I would be perfectly happy with declaring 0 to be invalid inside assembly names.)

I think that is mis-port. In native lexer 0 means an error in quoted state and end of input in unquoted. 

https://github.com/dotnet/runtime/blob/7f85a8c3887ba1be64b1eda4cc996fa543277848/src/coreclr/binder/stringlexer.cpp#L70

Reply:
It looks like there is enough inconsistencies between managed/native implementations that it is better be handled in a separate PR.

Reply:
> it is better be handled in a separate PR

I agree. It may make sense to add the test coverage and fix the inconsistencies as the first step so that the tests can give us confidence that the behavior is not changing with the switch.

Reply:
I think it might be easier to merge this PR and then fix corner cases and add tests. 

This way 
- we would not need to deal with differences in native implementations when adding tests as there will be only one implementation.
- we can still check preexisting behavior manually (simple app against 6.0).
- we would not need to deal with merges and conflicts of fixes in the managed lexer/parser against this change.

Reply:
> we would not need to deal with differences in native implementations when adding tests as there will be only one implementation.

The unmanaged parser is still used in number of cases. We want to keep them in sync unless you have a plan for deleting the unmanaged parser in near future.

Reply:
Once this API moves to managed. There will be the following entry points leading to assembly name parsing:

```c++
CorHost2::CreateDelegate(DWORD appDomainID, LPCWSTR wszAssemblyName, LPCWSTR wszClassName, LPCWSTR wszMethodName, INT_PTR* fnPtr);
MulticoreJitProfilePlayer::LoadAssembly(SString & assemblyName);
FriendAssemblyDescriptor::CreateFriendAssemblyDescriptor(PEAssembly *pAssembly);
LoadNativeLibraryBySearch(Assembly *callingAssembly, BOOL searchAssemblyDirectory, DWORD dllImportSearchPathFlags, LoadLibErrorTracker * pErrorTracker, LPCWSTR wszLibName);
DomainAssembly * LoadDomainAssembly(SString *  psszAssemblySpec, Assembly * pRequestingAssembly, AssemblyBinder * pBinder, BOOL       bThrowIfNotFound) ; // called by GetTypeWorker()
```

These call assembly spec `Init(LPSTR)`, which calls `ParseName`, which gets into assembly identity cache, which actually parses the name on cache misses.

I think removing the native assembly name parsing entirely could be done. 
We might need to put a cache on the managed side for that, considering we had a need for it in native.